### PR TITLE
support WindowsPrincipal in HttpListenerWebSocketContext

### LINKED ISSUE #63738: HttpListenerContext.AcceptWebSocketAsync unsupported for IWindowsPrincipal in .NET Core 3.1
### Description

HttpListenerContext.AcceptWebSocketAsync throws PlatformNotSupportedException when called with User set to IWindowsPrincipal. This happens in .NET Core 3.1 when running on Windows. Works fine when running in .NET Framework.

### Reproduction Steps

HttpListenerContext ctx = await listener.GetContextAsync();
wsCtx = await ctx.AcceptWebSocketAsync(WsConstants.SubProtocol, TimeSpan.FromSeconds(WsConstants.KeepAliveIntervalSeconds));

### Expected behavior

It should not throw the exception and work the same way as it does in .NET Framework

### Actual behavior

Throws PlatformNotSupportedException.

### Regression?

Yes, works in .NET Framework

### Known Workarounds

Use reflection to set the _user field to null, make the call and set it back to the original value, like this:

HttpListenerContext ctx = await listener.GetContextAsync();

IPrincipal originalPrincipal = ctx.User;
FieldInfo field = ctx.GetType().GetField("_user", BindingFlags.NonPublic | BindingFlags.Instance);
field.SetValue(_ctx, null);

// make the call with User = null to prevent the exception
wsCtx = await ctx.AcceptWebSocketAsync(WsConstants.SubProtocol, TimeSpan.FromSeconds(WsConstants.KeepAliveIntervalSeconds));

// revert to the original value
field.SetValue(_ctx, originalPrincipal);

### Configuration

.NET Core 3.1, Windows 10 x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #63738

### CODE REVIEW

File: src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs [hunk 1]
```diff
@@ -101,7 +101,15 @@ public class HttpListenerWebSocketContext : WebSocketContext
                 else
                 {
                     // AuthenticationSchemes.Digest, AuthenticationSchemes.Negotiate, AuthenticationSchemes.NTLM.
+#if (WINDOWS)
+                    if (user.Identity is WindowsIdentity windowsIdentity)
```
Comment (stephentoub) [inline src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs:105]: Don't we know user.Identity is a WindowsIdentity here?  The .NET Framework implementation just casts:
https://referencesource.microsoft.com/#System/net/System/Net/WebSockets/HttpListenerWebSocketContext.cs,149

File: src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs [hunk 2]
```diff
@@ -102,10 +102,8 @@ public class HttpListenerWebSocketContext : WebSocketContext
                 {
                     // AuthenticationSchemes.Digest, AuthenticationSchemes.Negotiate, AuthenticationSchemes.NTLM.
 #if (WINDOWS)
```
Comment (wfurt) [inline src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs:104]:
```suggestion
#if TARGET_WINDOWS
```

it seems we use that way in other places. (may impact cross-targeting and cross-compilation)

Comment (pedrobsaila) [reply src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs:104]: I tried it, but it seems unknown, here is a list of available windows preprocessor directives :
* WINDOWS
* WINDOWS7_0
* WINDOWS7_0_OR_GREATER

Comment (wfurt) [reply src/libraries/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs:104]: It seems to be set explicitly in project files that need it.
https://github.com/dotnet/runtime/blob/959a13fa0d1822dc4c7bfc99a9d5e7a0c50d6306/src/libraries/System.Net.Security/src/System.Net.Security.csproj#L13

**Review by wfurt — APPROVED** [review]
LGTM. test failures seems unrelated.

**Comment by MihaZupan** [comment]
Failures unrelated: #66672 #2271

**Pull request merged**